platform/linux-generic/odp_crypto_ipsecmb.c

```diff
@@ -0,0 +1,1063 @@
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	for (uint32_t i = 0; i < MAX_BUFFERS; i++) {
+		memcpy(local.pKeys[i], session->cipher.key_data,
+		session->p.cipher_key.length);
+		memcpy(local.pSrcData[i], data, in_len);
+		memcpy(local.pIV[i], iv_ptr, session->p.cipher_iv_len);
+		local.packetLen[i] = param->cipher_range.length;
+	}
+
+	IMB_ZUC_EEA3_N_BUFFER(mb_mgr, (const void * const *)local.pKeys,
+			      (const void * const *)local.pIV,
+			      (const void * const *)local.pSrcData,
+			      (void **)local.pDstData, local.packetLen,
+			      MAX_BUFFERS);
+
+	odp_packet_copy_from_mem(pkt, in_pos, in_len, local.pDstData[0]);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static
+void zuc_eea3_decrypt(odp_packet_t pkt,
+		      const odp_crypto_packet_op_param_t *param,
+		      odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+
+#if ODP_DEPRECATED_API
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else {
+		set_crypto_op_result(pkt,
+				     ODP_CRYPTO_ALG_ERR_DATA_SIZE,
+				     ODP_CRYPTO_ALG_ERR_NONE);
+		return;
+	}
+#else
+	iv_ptr = param->cipher_iv_ptr;
+	_ODP_ASSERT(session->p.cipher_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(odp_packet_is_segmented(pkt))) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	for (uint32_t i = 0; i < MAX_BUFFERS; i++) {
+		memcpy(local.pKeys[i], session->cipher.key_data,
+			session->p.cipher_key.length);
+		memcpy(local.pSrcData[i], data, in_len);
+		memcpy(local.pIV[i], iv_ptr, session->p.cipher_iv_len);
+		local.packetLen[i] = in_len;
+	}
+
+	IMB_ZUC_EEA3_N_BUFFER(mb_mgr, (const void * const *)local.pKeys,
+			      (const void * const *)local.pIV,
+			      (const void * const *)local.pSrcData,
+			      (void **)local.pDstData, local.packetLen,
+			      MAX_BUFFERS);
+
+	odp_packet_copy_from_mem(pkt, in_pos, in_len, local.pDstData[0]);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static int process_zuc_eea3_param(odp_crypto_generic_session_t *session)
+{
+	/* Verify key length is valid */
+	if (16 != session->p.cipher_key.length)
+		return -1;
+
+	/* Verify the IV length is correct */
+	if (16 != session->p.cipher_iv_len)
+		return -1;
+
+	memcpy(session->cipher.key_data, session->p.cipher_key.data,
+		session->p.cipher_key.length);
+
+	/* Set function */
+	if (ODP_CRYPTO_OP_ENCODE == session->p.op)
+		session->func = zuc_eea3_encrypt;
+	else
+		session->func = zuc_eea3_decrypt;
```

Comment:
As mentioned earlier, this is buggy. You have single func pointer in the session but you are trying to use it for both the cipher and authentication function. One or the other gets overwritten depending on the order in which the process_xxx_param() gets called. ODP API allows both a ciphering and an authentication algorithm to be applied to a packet in single crypto operation, so this code should be able to handle it. Also the order in which ciphering and authentication is done is configurable, which you have to take in to account.

Reply:
Thanks for your comments, added separated function callback for cipher and authenticate propose. Added `session->do_cipher_first` to handle cipher and auth orders.